src/model/wallet.js

```diff
@@ -831,3 +805,124 @@ export const loadSeedPhraseAndMigrateIfNeeded = async id => {
     return null;
   }
 };
+
+async function extractSecretsForWallet(wallet) {
+  const allKeys = await keychain.loadAllKeys();
+  if (!allKeys) throw new Error("Couldn't read secrets from keychain");
+  const secrets = {};
+
+  const allowedPkeysKeys = wallet.addresses.map(
+    account => `${account.address}_${privateKeyKey}`
+  );
+
+  allKeys.forEach(item => {
+    // Ignore allWalletsKey
+    if (item.username === allWalletsKey) {
+      return;
+    }
+
+    // Ignore selected wallet
+    if (item.username === selectedWalletKey) {
+      return;
+    }
+
+    // Ignore icloud backup password
+    if (item.username === 'rainbowBackup') {
+      return;
+    }
+
+    // Ignore another wallets seeds
+    if (
+      item.username.indexOf(`_${seedPhraseKey}`) !== -1 &&
+      item.username !== `${wallet.id}_${seedPhraseKey}`
+    ) {
+      return;
+    }
+
+    // Ignore other wallets PKeys
+    if (
+      item.username.indexOf(privateKeyKey) !== -1 &&
+      allowedPkeysKeys.indexOf(item.username) === -1
+    ) {
+      return;
+    }
+
+    secrets[item.username] = item.password;
+  });
+  return secrets;
+}
+
+export async function backupWalletToCloud(password, wallet) {
+  const now = Date.now();
+
+  const secrets = await extractSecretsForWallet(wallet);
+  const data = {
+    createdAt: now,
+    secrets,
+  };
+  return encryptAndSaveDataToCloud(data, password, `backup_${now}.json`);
+}
+
+export async function addWalletToCloudBackup(password, wallet, filename) {
+  const backup = await getDataFromCloud(password, filename);
```

Comment:
Just adding another comment here as my response may have gotten lost - basically let's do a null check and fail early if backup is null (like how we are in line 911 but we can return null instead of throwing)